From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 8 Mar 2022 00:37:35 +1000
Subject: [PATCH] Only drop leads on the local server


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index e59bc3e620716beae1a7aed9b212721d7f66bda3..f68e6797a5b96ffb773200c4f61cc773a20f050e 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -86,6 +86,7 @@ import org.bukkit.event.entity.EntityTransformEvent;
 import org.bukkit.event.entity.EntityUnleashEvent;
 import org.bukkit.event.entity.EntityUnleashEvent.UnleashReason;
 // CraftBukkit end
+import puregero.multipaper.MultiPaper; // MultiPaper
 
 public abstract class Mob extends LivingEntity implements Targeting {
 
@@ -1527,7 +1528,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
         if (this.leashHolder != null) {
             this.leashHolder = null;
             this.leashInfoTag = null;
-            if (!this.level.isClientSide && dropItem) {
+            if (!this.level.isClientSide && dropItem && MultiPaper.isChunkLocal(level.getChunkIfLoaded(blockPosition()))) { // MultiPaper - only drop leads on the local server
                 this.forceDrops = true; // CraftBukkit
                 this.spawnAtLocation((ItemLike) Items.LEAD);
                 this.forceDrops = false; // CraftBukkit
