From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 5 Jul 2022 17:48:07 +1000
Subject: [PATCH] Only wake up local players


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index a83222b83dab679b6729f83589b7fe13487e4477..d2886a68d13aa6a59bfac5ca31f07f5be5d09c49 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -842,7 +842,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
     private void wakeUpAllPlayers() {
         this.sleepStatus.removeAllSleepers();
-        (this.players.stream().filter(LivingEntity::isSleeping).collect(Collectors.toList())).forEach((entityplayer) -> { // CraftBukkit - decompile error
+        (this.players.stream().filter(MultiPaper::isRealPlayer).filter(LivingEntity::isSleeping).collect(Collectors.toList())).forEach((entityplayer) -> { // CraftBukkit - decompile error // MultiPaper - only wake up local players
             entityplayer.stopSleepInBed(false, false);
         });
     }
