From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sun, 29 Jan 2023 16:33:09 +0100
Subject: [PATCH] Fix Ikea V1 dupe


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index dd830d509c83cb44ccf87ddd2c3fed55826235c8..0ff0f2b4018621b687c91126cae8b22fc5f371a5 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3035,6 +3035,12 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                 }).collect(ImmutableList.toImmutableList());
             }
 
+            // MultiPaper start
+            if (entity instanceof ServerPlayer player) {
+                player.closeContainer();
+            }
+            // MultiPaper end
+
             entity.boardingCooldown = 60;
             this.gameEvent(GameEvent.ENTITY_DISMOUNT, entity);
         }
